torch/csrc/autograd/function.h

```diff
@@ -90,9 +133,24 @@ struct Function : std::enable_shared_from_this<Function> {
   };
 
   // Computes is_executable and next_functions from a list of input variables
-  static FunctionFlags flags(const variable_list& inputs);
-  static FunctionFlags flags(const std::initializer_list<Variable>& inputs);
-  static FunctionFlags flags(at::TensorList inputs);
+  // (but which are tensors, because this is what happens in the autogenerated
+  // code)
+  template<typename... Args> inline static FunctionFlags tensor_flags(Args&&... args) {
+    FunctionFlags f;
+    if (!GradMode::is_enabled()) return f;
+    f.next_functions.reserve(count_tensors(std::forward<Args>(args)...));
+    SetTensorFunctionFlags(f).apply(std::forward<Args>(args)...);
+    return f; // RVO
+  }
+
+  // Computes is_executable and next_functions from a list of input variables
+  template<typename... Args> inline static FunctionFlags flags(Args&&... args) {
```

Comment:
Can we finally remove `requires_grad` from these flags, and assume that if you call it, then you really want to get a list of next functions? We're doing a bunch of repeated checks for GradMode in different places now.

Reply:
Also: this doesn't only handle Variables - it also handles vectors/array refs of Variables.

Reply:
This seems orthogonal to this patchset. I am also not convinced that we can assume that the precondition you suggested holds for all call sites of `flags`. For example, it's called from `wrap_outputs` and I'm pretty sure `wrap_outputs` is always called no matter if we require grads or not.

I fixed the comment though.